string: implement API to access raw string data

indygreg opened the pull request:
With the recent implementation of non-limited unicode APIs, we're
able to query Python's low-level state to access the raw bytes that
Python is using to store string objects.

This commit implements a safe Rust API for obtaining a view into
Python's internals and representing the raw bytes Python is using
to store strings.

Not only do we allow accessing what Python has stored internally,
but we also support coercing this data to a `Cow<str>`.

Closes #1776.

davidhewitt commented on CHANGELOG.md:
```diff
@@ -8,6 +8,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
+### Added
+
+- Implement `PyString.data()` to access the raw bytes storing a Python string. [#1776](https://github.com/PyO3/pyo3/issues/1776)
```
The convention in PyO3 has been to attach the PR number to the CHANGELOG, with the thinking that users can start there and read backwards through the conversation.

```suggestion
- Implement `PyString.data()` to access the raw bytes storing a Python string. [#1794](https://github.com/PyO3/pyo3/issues/1794)
```

davidhewitt reviewed (commented):
Thanks, this looks useful to me! I wonder whether eventually we'd use this internally in PyO3 too to potentially avoid causing strings to generate (and cache) their UTF-8 representation.

I have a bunch of small nitty comments below.

I'd also very much like to see some more tests of `PyStringData::to_string()`  - it would be good to cover both the success case (which looks like it's covered) and the error cases (which don't appear to be) for each of the three character widths.

davidhewitt commented:
Also, `types::string` is currently not a public module, which is why CI is unhappy. I don't personally see the harm in making it public? I think I'd rather users wrote `use pyo3::types::string::PyStringData` over `use pyo3::types:::PyStringData` - the latter seems misleading.

indygreg commented on src/types/string.rs:
```diff
@@ -89,6 +199,48 @@ impl PyString {
+            ffi::PyUnicode_4BYTE_KIND => Ok(PyStringData::Ucs4(unsafe {
+                std::slice::from_raw_parts(raw_data as *const u32, length)
+            })),
+            kind => Err(crate::exceptions::PyValueError::new_err(format!(
+                "unknown string kind: {}",
+                kind
+            ))),
```
I'm OK with it if you are. I usually try to avoid panics if possible. But as long as we stay on top of API contracts, this should be fine.

indygreg commented:
OK. I think the most recent push addresses most (all?) of your feedback.

I did disagree with you on exposing `PyStringData` under `pyo3::types::string`. If you insist on that, suggest a commit and I'll apply it.

I added tests for verifying `.to_string()` errors. This required using `PyUnicode_FromKindAndData()` to construct instances with invalid byte sequences, which Python appears to accept without validation.

As part of writing the tests, I noticed there may be a `c_int` / `c_uint` mismatch in the ffi definitions. When I implemented the FFI bindings, I attempted to copy signedness from the C headers. However, some of these things were macros and lacked type annotations. I may have used the wrong types for symbols like `PyUnicode_1BYTE_KIND` :/

indygreg commented on src/types/string.rs:
```diff
@@ -1,14 +1,129 @@
+                    )?))
+                }
+            },
+            Self::Ucs4(data) => match data.iter().map(|&c| char::from_u32(c)).collect() {
```
`std::char::from_u32` and the corresponding `TryFrom<u32>` conversion appear to be added in Rust 1.34. The implementation literally just checks for `if (i > MAX as u32) || (i >= 0xD800 && i <= 0xDFFF)` so it would be trivial to inline. I suppose we'll need to do that to support Rust 1.32.

Unless you have any better suggestions...

birkenfeld replied:
PyO3 MSRV is 1.41 though? What am I missing?

indygreg replied:
Err. You are right. CI was complaining when I used `char::from_u32`.

It appears `char::from_u32` and `std::char::from_u32` are different. The former works on stable but not on 1.41. TIL. I'll push a fix.

birkenfeld commented on src/types/string.rs:
```diff
@@ -89,6 +199,48 @@ impl PyString {
+    ///
+    /// If the Python string object was created through legacy APIs, its internal
+    /// storage format will be canonicalized before data is returned.
+    #[cfg(not(Py_LIMITED_API))]
```
is this available on PyPy?

davidhewitt replied:
Looks like the PyPy CI is working fine!

birkenfeld commented on src/types/string.rs:
```diff
@@ -1,14 +1,129 @@
+    /// returns an owned string.
+    ///
+    /// Returns [PyUnicodeDecodeError] if the string data isn't valid in its purported.
+    #[allow(clippy::wrong_self_convention)]
```
Since this struct is `Copy`, why can't you take `self` instead of silencing the lint?

indygreg replied:
Good suggestion!

birkenfeld commented on src/types/string.rs:
```diff
@@ -1,14 +1,129 @@
+        }
+    }
+
+    /// Size in bytes of each *character* in the underlying buffer.
```
Is "character" the correct terminology here? IMO it should be "codepoint".

indygreg replied:
It is debatable.

It is unclear to me if Python's internal storage format actually guarantees that a slice N byte value is the same as a _code point_. UTF-8 and UTF-16 are variable width encodings after all. I _think_ Python will do things like store surrogate pairs without expanding the item width of the stored value.

The documentation at https://docs.python.org/3/c-api/unicode.html uses _character_ to refer to the values in these buffers, which is the main reason I used _character_ in the API.

But, I don't likely blindly copying things, especially if they are wrong.

What I know for certain is that the values stored in the memory slice are stored with a fixed width. So maybe we should go with a name that says what they actually are: `item width in bytes`.

🤷‍♂️

birkenfeld commented on src/types/string.rs:
```diff
@@ -1,14 +1,129 @@
+    /// For UCS-1 / UTF-8, returns a borrow into the original slice. For UCS-2 and UCS-4,
+    /// returns an owned string.
+    ///
+    /// Returns [PyUnicodeDecodeError] if the string data isn't valid in its purported.
```
IIUC exceptions can only happen if strings with invalid encoding are created using the C API, right? Would be nice to clarify that (also in to_string_lossy below).

indygreg replied:
I'll update the docs in the next push.

davidhewitt reviewed (approved):
Nice work, this looks good to me!
